### ISSUE TITLE
rasterio ValueError: negative dimensions are not allowed when dask_chunks is set

### ISSUE THREAD

**Issue by loicdtx** [opendatacube/datacube-core#351]
Hi,

I'm running tests on a local datacube deployment in which we've ingested Landsat 8 SR data over a state of Mexico.
I'm trying to parallelize simple computation with dask using the `dask_chunk` argument of the `load()` method, but I'm getting the following error:

```
Error opening source dataset: NetCDF:/LUSTRE/MADMEX/datacube_ingest/LS8_espa/mexico/LS8_espa_14_-15_20170105171146381304.nc:swir2
Traceback (most recent call last):
  File "/home/madmex_user/sandbox/dask_dc_test.py", line 16, in <module>
    print(sr_mean.load())
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/xarray/core/dataset.py", line 491, in load
    evaluated_data = da.compute(*lazy_data.values(), **kwargs)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/base.py", line 333, in compute
    results = get(dsk, keys, **kwargs)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 562, in get_sync
    return get_async(apply_sync, 1, dsk, keys, **kwargs)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 508, in get_async
    fire_task()
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 504, in fire_task
    callback=queue.put)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 551, in apply_sync
    res = func(*args, **kwds)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 295, in execute_task
    result = pack_exception(e, dumps)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 290, in execute_task
    result = _execute_task(task, data)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 270, in _execute_task
    args2 = [_execute_task(a, cache) for a in args]
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 270, in <listcomp>
    args2 = [_execute_task(a, cache) for a in args]
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/dask/local.py", line 271, in _execute_task
    return func(*args2)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/datacube/api/core.py", line 679, in fuse_lazy
    _fuse_measurement(data, datasets, geobox, measurement, fuse_func=fuse_func, driver_manager=driver_manager)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/datacube/api/core.py", line 692, in _fuse_measurement
    skip_broken_datasets=skip_broken_datasets)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/datacube/storage/storage.py", line 205, in reproject_and_fuse
    read_from_source(source, buffer_, dst_transform, dst_nodata, dst_projection, resampling)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/datacube/storage/storage.py", line 152, in read_from_source
    tmp, offset, _ = _read_decimated(array_transform, src, dest.shape)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/datacube/storage/storage.py", line 121, in _read_decimated
    tmp = src.read(window=window, out_shape=write_shape)
  File "/home/madmex_user/.virtualenvs/antares/lib/python3.5/site-packages/datacube/storage/storage.py", line 243, in read
    return self.source.ds.read(indexes=self.source.bidx, window=window, out_shape=out_shape)
  File "rasterio/_io.pyx", line 299, in rasterio._io.DatasetReaderBase.read
ValueError: negative dimensions are not allowed
```

Interestingly things work as expected if I process in memory, without the dask option.

### The code used to reproduce the behaviour:

```python
import datacube
from datacube.storage.storage import write_dataset_to_netcdf
from datetime import datetime

import dask
import dask.multiprocessing
dask.set_options(get=dask.multiprocessing.get)

dc = datacube.Datacube(app = 'dask_test')
sr = dc.load(product='ls8_espa_mexico', x=(-105, -102), y=(19, 22),
                      time=(datetime(2017, 1, 1), datetime(2018, 1, 1)), group_by='solar_day',
                      dask_chunks={'x': 2000, 'y': 2000}, use_threads=True)
print(sr)
# <xarray.Dataset>
# Dimensions:   (time: 69, x: 10501, y: 11196)
# Coordinates:
#   * time      (time) datetime64[ns] 2017-01-03T17:24:08.383887 ...
#   * y         (y) float64 1.119e+06 1.119e+06 1.119e+06 1.119e+06 1.119e+06 ...
#   * x         (x) float64 2.185e+06 2.185e+06 2.185e+06 2.185e+06 2.185e+06 ...
# Data variables:
#     blue      (time, y, x) int16 dask.array<shape=(69, 11196, 10501), chunksize=(69, 2000, 2000)>
#     green     (time, y, x) int16 dask.array<shape=(69, 11196, 10501), chunksize=(69, 2000, 2000)>
#     red       (time, y, x) int16 dask.array<shape=(69, 11196, 10501), chunksize=(69, 2000, 2000)>
#     nir       (time, y, x) int16 dask.array<shape=(69, 11196, 10501), chunksize=(69, 2000, 2000)>
#     swir1     (time, y, x) int16 dask.array<shape=(69, 11196, 10501), chunksize=(69, 2000, 2000)>
#     swir2     (time, y, x) int16 dask.array<shape=(69, 11196, 10501), chunksize=(69, 2000, 2000)>
#     pixel_qa  (time, y, x) int16 dask.array<shape=(69, 11196, 10501), chunksize=(69, 2000, 2000)>
# Attributes:
#     crs:      PROJCS["unnamed",GEOGCS["WGS 84",DATUM["unknown",SPHEROID["WGS8...

sr_mean = sr.mean('time')
print(sr_mean.load())
```

### Environment information

```
datacube --version
Open Data Cube core, version 1.5.1+249.g98bc783
```

**Comment by andrewdhicks** [comment]
Hi @loicdtx,

I think it is due to a bug where we try to access a chunked window that doesn't overlap with the file. We don't check to see if the file is actually needed for the chunk and try to load it anyway.

It looks like we fixed the bug in version 1.5.2, so if you can try a newer version, hopefully it should work! If it doesn't, comment here and we'll re-open then issue.

The `dask` support hasn't been widely used, so if you have any issues with it, let us know here or on Slack. We'd also love to hear how you are using Open Data Cube!

Thanks,
Andrew

**Comment by loicdtx** [comment]
Thanks for the prompt response @andrewdhicks 
I actually installed from the HEAD of the develop branch, so I should already be on the most up to date version. Or does the `release-1.5` contain fixes that haven't been merged in `develop`?

Here's how I installed it.

```bash
mktmpenv
pip install numpy
pip install GDAL==$(gdal-config --version) --global-option=build_ext --global-option="-I/usr/include/gdal"
cd /tmp
git clone https://github.com/opendatacube/datacube-core.git
cd datacube-core
pip install -e .
datacube --version
# Open Data Cube core, version 1.5.1+253.g792c6d9
```

Feel free to message me via the opendatacube slack group if you want to discuss the issue more interactively.

**Comment by omad** [comment]
Thanks for the detailed bug report! It looks very similar to a problem @petewa sent me last year, which included a fix which I've just pushed to the `origin/fix-negative-window` branch. If you were able to run the same test against this and let us know how you go.

There are unfortunately some fixes in `release-1.5` which haven't been merged into `develop` yet. We are working on doing this ASAP.

**Comment by loicdtx** [comment]
Yes, I saw yesterday that the two branches are quite different. What I did was switching to `release-1.5` and cherry picking the features of `develop` I was interested in.
I installed from [here](https://github.com/CONABIO/datacube-core/tree/release-1.5) using `pip install git+https://github.com/CONABIO/datacube-core.git@release-1.5`
Everything's working now! Thanks!
